pkg/distributor/errors.go

```diff
@@ -158,6 +159,37 @@ func (e requestRateLimitedError) errorCause() mimirpb.ErrorCause {
 // Ensure that requestRateLimitedError implements distributorError.
 var _ distributorError = requestRateLimitedError{}
 
+// ingesterPushError is an error used to represent a failed attempt to push to the ingester.
+type ingesterPushError struct {
+	message string
+	cause   mimirpb.ErrorCause
+}
+
+// newIngesterPushError creates a ingesterPushError error representing the given status object.
+func newIngesterPushError(stat *status.Status) ingesterPushError {
+	errorCause := mimirpb.INVALID
+	if len(stat.Details()) == 1 {
+		if errorDetails, ok := stat.Details()[0].(*mimirpb.WriteErrorDetails); ok {
```

Comment:
since `.Details()` performs unmarshaling and it allocates a slice for the unmarshaled details, should we re-use its results instead of calling it twice?

Reply:
I agree, it would also be clearer in the code that we're checking the same details slice (who knows if `Details()` may return different length results)

Reply:
A great catch. Thank you. I will fix this